src/main/java/org/opensearch/knn/index/query/KNNQueryBuilder.java

```diff
@@ -341,36 +343,49 @@ protected Query doToQuery(QueryShardContext context) {
         if (!(mappedFieldType instanceof KNNVectorFieldType)) {
             throw new IllegalArgumentException(String.format(Locale.ROOT, "Field '%s' is not knn_vector type.", this.fieldName));
         }
-
         KNNVectorFieldType knnVectorFieldType = (KNNVectorFieldType) mappedFieldType;
-        int fieldDimension = knnVectorFieldType.getDimension();
-        KNNMethodContext knnMethodContext = knnVectorFieldType.getKnnMethodContext();
-        MethodComponentContext methodComponentContext = null;
-        KNNEngine knnEngine = KNNEngine.DEFAULT;
-        VectorDataType vectorDataType = knnVectorFieldType.getVectorDataType();
-        SpaceType spaceType = knnVectorFieldType.getSpaceType();
-        VectorQueryType vectorQueryType = getVectorQueryType(k, maxDistance, minScore);
-        updateQueryStats(vectorQueryType);
-
-        if (fieldDimension == -1) {
-            if (spaceType != null) {
-                throw new IllegalStateException("Space type should be null when the field uses a model");
-            }
-            // If dimension is not set, the field uses a model and the information needs to be retrieved from there
-            ModelMetadata modelMetadata = getModelMetadataForField(knnVectorFieldType);
+        ANNConfig annConfig = knnVectorFieldType.getAnnConfig();
+        if (ANNConfigType.SKIP == annConfig.getAnnConfigType()) {
+            throw new IllegalArgumentException(String.format(Locale.ROOT, "Field '%s' is not built for ANN search.", this.fieldName));
+        }
+
+        int fieldDimension;
+        KNNEngine knnEngine;
+        MethodComponentContext methodComponentContext;
+        SpaceType spaceType;
+        VectorDataType vectorDataType;
+        if (ANNConfigType.FROM_MODEL == annConfig.getAnnConfigType()) {
```

Comment:
Feels like all of these can be encapsulated inside ANNConfigType class and avoid if/else here if we maintain different AnnConfigType for each type.

Reply:
Sure, Im going to get rid of the AnnConfigType and implement inside mappers